app/code/Magento/Catalog/Block/Product/ProductList/Upsell.php

```diff
@@ -262,11 +267,10 @@ public function getItemLimit($type = '')
      */
     public function getIdentities()
     {
-        $identities = [];
-        foreach ($this->getItems() as $item) {
-            // phpcs:ignore Magento2.Performance.ForeachArrayMerge
-            $identities = array_merge($identities, $item->getIdentities());
-        }
-        return $identities;
+        $identities = array_map(function (DataObject $item) {
+            return $item->getIdentities();
+        }, $this->getItems());
+
+        return array_merge(...$identities);
```

Comment:
I think there might be an issue when `$this->getItems()` will return empty array

Reply:
Let me cover that case with unit tests

Reply:
I wanted to write Unit Tests the right way, but I'll use the existing way to mock `items` - sooner or later, both blocks have to be rewritten testable-way.

Reply:
You were right but not about `getItems`. Unit test caught it and I fixed this with fallback to `[[]]`